pkg/daemon/daemon.go

```diff
@@ -924,10 +803,55 @@ func (dn *Daemon) drainNode() error {
 		glog.Errorf("drainNode(): failed to drain node: %v", err)
 		return err
 	}
+	dn.enableUpdatingFlag()
 	glog.Info("drainNode(): drain complete")
 	return nil
 }
 
+func (dn *Daemon) getUpdatingFlag() bool {
+	dn.updatingFlagMutex.Lock()
+	defer dn.updatingFlagMutex.Unlock()
+	return dn.updatingFlag
+}
+
+func (dn *Daemon) enableUpdatingFlag() {
+	dn.updatingFlagMutex.Lock()
+	defer dn.updatingFlagMutex.Unlock()
+	glog.Info("enableUpdatingFlag(): Setting flag to true")
+	dn.updatingFlag = true
+}
+
+func (dn *Daemon) disableUpdatingFlag() {
+	dn.updatingFlagMutex.Lock()
+	defer dn.updatingFlagMutex.Unlock()
+	glog.Info("disableUpdatingFlag(): Setting flag to false")
+	dn.updatingFlag = false
+}
+
+func (dn *Daemon) installSignalHandler(stopCh chan struct{}) {
+	termChan := make(chan os.Signal, 2048)
+	signal.Notify(termChan, syscall.SIGTERM)
+
+	// Catch SIGTERM - if we're actively updating, we should avoid
+	// having the process be killed.
+	go func() {
+		for sig := range termChan {
+			//nolint:gocritic
+			switch sig {
+			case syscall.SIGTERM:
+				updateActive := dn.getUpdatingFlag()
+				if updateActive {
+					glog.Info("Got SIGTERM, but actively updating")
```

Comment:
So what happens with the SIGTERM in this case? We are already updating in the sriov daemon so we just ignore the request to shut down since we will end our updating activity with a reboot? What if we require a drain but not a reboot? If reqDrain, then we call drainNode and we enable the flag to ignore the signal. Is it the case that in such scenario, we were asked to shutdown, but we ignored that completely?

Reply:
The way this works is:
- Someone issues a reboot/shutdown.
- Kubelet gets the event and creates two lists of pods: critical and non critical, based on priorityClass.
- Starting with non-critical pods, it checks the minimum between the termination period for the pod and the configured timeout in kubelet.
- Send SIGTERM using eviction API to pod. Check for pod termination for the computed value in the previous bullet. If the pod is still alive, send SIGKILL.
- Repeat operation with critical pods.
- Shutdown.

If a drain must happen then SIGTERM is ignored while doing so. The timeout for termination must be properly configured to allow enough time to completion. After that timeout a SIGKILL will end the process.

If a reboot is required then its the same situation but in other pods. The daemon will issue a shutdown that will kickstart this process in the node. In the end everyone will receive a SIGTERM followed by a SIGKILL if the pod is still alive after the termination grace period.

If there is no need to drain/reboot then SIGTERM will end the process.

Once SIGTERM has been received it marks that the pod is bound to terminate some time in the near future.

Reply:
Thank you for clarifying. Please add this explanation to the commit message for future reference too.

Reply:
Added in the first commit.